interchaintest/module_clock_test.go

```diff
@@ -33,40 +33,131 @@ func TestJunoClock(t *testing.T) {
 	users := interchaintest.GetAndFundTestUsers(t, ctx, "default", int64(10_000_000_000), juno, juno)
 	user := users[0]
 
-	// Upload & init contract payment to another address
+	//
+	// -- REGULAR GAS CONTRACT --
+	// Ensure logic works as expected for a contract that uses less than the gas limit
+	// and has a valid sudo message entry point.
+	//
+
+	// Setup contract
 	_, contractAddr := helpers.SetupContract(t, ctx, juno, user.KeyName(), "contracts/clock_example.wasm", `{}`)
 
 	// Ensure config is 0
 	res := helpers.GetClockContractValue(t, ctx, juno, contractAddr)
 	fmt.Printf("- res: %v\n", res.Data.Val)
 	require.Equal(t, uint32(0), res.Data.Val)
 
-	// Submit the proposal to add it to the allowed contracts list
-	SubmitParamChangeProp(t, ctx, juno, user, []string{contractAddr})
+	// Register the contract
+	helpers.RegisterClockContract(t, ctx, juno, user, contractAddr)
 
 	// Wait 1 block
 	_ = testutil.WaitForBlocks(ctx, 1, juno)
 
+	// Validate contract is not jailed
+	contract := helpers.GetClockContract(t, ctx, juno, contractAddr)
+	require.False(t, contract.ClockContract.IsJailed)
+
 	// Validate the contract is now auto incrementing from the end blocker
 	res = helpers.GetClockContractValue(t, ctx, juno, contractAddr)
 	fmt.Printf("- res: %v\n", res.Data.Val)
 	require.GreaterOrEqual(t, res.Data.Val, uint32(1))
 
+	// Unregister the contract & ensure it is removed from the store
+	helpers.UnregisterClockContract(t, ctx, juno, user, contractAddr)
+	helpers.ValidateNoClockContract(t, ctx, juno, contractAddr)
+
+	//
+	// -- HIGH GAS CONTRACT --
+	// Ensure contracts that exceed the gas limit are jailed.
+	//
+
+	// Setup contract
+	_, contractAddr = helpers.SetupContractWithAdmin(t, ctx, juno, user.KeyName(), user.FormattedAddress(), "contracts/clock_example_high_gas.wasm", `{}`)
+
+	// Ensure config is 0
+	res = helpers.GetClockContractValue(t, ctx, juno, contractAddr)
+	fmt.Printf("- res: %v\n", res.Data.Val)
+	require.Equal(t, uint32(0), res.Data.Val)
+
+	// Register the contract
+	helpers.RegisterClockContract(t, ctx, juno, user, contractAddr)
+
+	// Wait 1 block
+	_ = testutil.WaitForBlocks(ctx, 1, juno)
+
+	// Validate contract is jailed
+	contract = helpers.GetClockContract(t, ctx, juno, contractAddr)
+	require.True(t, contract.ClockContract.IsJailed)
+
+	//
+	// -- MIGRATE CONTRACT --
+	// Ensure migrations can patch contracts that error or exceed gas limit
+	// so they can be unjailed.
+	//
+
+	// Migrate the high gas contract to a contract with lower gas usage
+	helpers.MigrateContract(t, ctx, juno, user.KeyName(), contractAddr, "contracts/clock_example_migrate.wasm", `{}`)
+
+	// Unjail the contract
+	helpers.UnjailClockContract(t, ctx, juno, user, contractAddr)
```

Comment:
If the contract unjails but did not migrate, the contract will remain jailed correct after the next block?
(So the flow is contractJailed -> user unjails -> EndBlock -> contractFailsAgain -> ReJail)

Reply:
Yes, that is correct. If the contract is not patched, it will continually get re-jailed over and over again.